src/main/scala/is/hail/table/Table.scala

```diff
@@ -1138,4 +1139,95 @@ class Table(val hc: HailContext, val tir: TableIR) {
+    assert(keySignature.exists(s => s.size == 1 && s.types(0) == intervalType.pointType))
+
+    val leftORVD = rvd match {
+      case ordered: OrderedRVD => ordered
+      case unordered =>
+        OrderedRVD.coerce(
+          new OrderedRVDType(key.get.toArray, key.get.toArray, signature),
+          unordered)
+    }
+
+
+    val typOrdering = intervalType.pointType.ordering
+
+    val typToInsert: Type = other.valueSignature
+
+    val (newRVType, ins) = signature.unsafeStructInsert(typToInsert, List(fieldName))
+
+    val partBc = hc.sc.broadcast(leftORVD.partitioner)
+    val rightSignature = other.signature
+    val rightKeyFieldIdx = other.keyFieldIdx.get(0)
+    val rightValueFieldIdx = other.valueFieldIdx
+    val partitionKeyedIntervals = other.rvd.boundary.crdd
+      .cflatMap { (ctx, rv) =>
+        val region = ctx.region
+        val r = SafeRow(rightSignature, rv)
+        val ur = new UnsafeRow(rightSignature, rv)
+        val interval = ur.getAs[Interval](rightKeyFieldIdx)
+        if (interval != null) {
+          val rangeTree = partBc.value.rangeTree
+          val pkOrd = partBc.value.pkType.ordering
+          val wrappedInterval = interval.copy(
+            start = Row(Annotation.copy(intervalType.pointType, interval.start)),
+            end = Row(Annotation.copy(intervalType.pointType, interval.end)))
+          rangeTree.queryOverlappingValues(pkOrd, wrappedInterval).map(i => (i, r))
+        } else
+          Iterator()
+      }
+
+    val nParts = rvd.getNumPartitions
+    val zipRDD = partitionKeyedIntervals.partitionBy(new Partitioner {
+      def getPartition(key: Any): Int = key.asInstanceOf[Int]
+
+      def numPartitions: Int = nParts
+    }).values
+
+    val localRVRowType = signature
+    val pkIndex = signature.fieldIdx(key.get(0))
+    val newTableType = typ.copy(rowType = newRVType)
+    val newOrderedRVType = new OrderedRVDType(key.get.toArray, key.get.toArray, newRVType)
+    val newRVD = leftORVD.zipPartitionsPreservesPartitioning(
+      newOrderedRVType,
+      zipRDD
+    ) { case (it, intervals) =>
+      val intervalAnnotations: Array[(Interval, Any)] =
+        intervals.map { r =>
+          val interval = r.getAs[Interval](rightKeyFieldIdx)
+          (interval, Row.fromSeq(rightValueFieldIdx.map(r.get)))
+        }.toArray
+
+      val iTree = IntervalTree.annotationTree(typOrdering, intervalAnnotations)
+
+      val rvb = new RegionValueBuilder()
+      val rv2 = RegionValue()
+
+      it.map { rv =>
+        val ur = new UnsafeRow(localRVRowType, rv)
+        val pk = ur.get(pkIndex)
+        val queries = iTree.queryValues(typOrdering, pk)
+        val value = if (queries.isEmpty)
+          null
+        else
+          queries(0)
+        assert(typToInsert.typeCheck(value))
+
+        rvb.set(rv.region)
+        rvb.start(newRVType)
+
+        ins(rv.region, rv.offset, rvb, () => rvb.addAnnotation(typToInsert, value))
```

Comment:
Hmm. This seems like it'd be faster if interval tree's contained encoded bytes rather than `Annotation`. Then we'd just be decoding into the region.

Reply:
Hmm, I don't think the decoder can use an `rvb` though, so this might be some non-trivial work to make decoders know how to use rvbs.